src/workerd/api/basics.c++

```diff
@@ -669,6 +706,107 @@ void AbortSignal::triggerAbort(
+    serializer.write(js, r.getHandle(js));
+  } else {
+    serializer.write(js, js.undefined());
+  }
+
+  auto streamCap = externalHandler
+                       ->writeStream([&](rpc::JsValue::External::Builder builder) mutable {
+    builder.setAbortSignal();
+  }).castAs<rpc::AbortSignal>();
+
+  auto& ioContext = IoContext::current();
+  // Keep track of every AbortSignal cloned from this one.
+  // If this->triggerAbort(...) is called, each rpcClient will be informed.
+  rpcClients.add(ioContext.addObject(kj::heap(kj::mv(streamCap))));
+}
+
+jsg::Ref<AbortSignal> AbortSignal::deserialize(
+    jsg::Lock& js, rpc::SerializationTag tag, jsg::Deserializer& deserializer) {
+  auto& handler = KJ_REQUIRE_NONNULL(
+      deserializer.getExternalHandler(), "got AbortSignal on non-RPC serialized object?");
+  auto externalHandler = dynamic_cast<RpcDeserializerExternalHander*>(&handler);
+  KJ_REQUIRE(externalHandler != nullptr, "got AbortSignal on non-RPC serialized object?");
+
+  auto isCanceled = static_cast<bool>(deserializer.readRawUint32());
+  auto flag = static_cast<Flag>(deserializer.readRawUint32());
+  auto reason = deserializer.readValue(js);
+
+  auto reader = externalHandler->read();
+  KJ_REQUIRE(reader.isAbortSignal(), "external table slot type does't match serialization tag");
+
+  if (isCanceled) {
+    // The signal is already aborted and cannot be triggered again. We don't need to set up RPC.
+    return abort(js, reason);
+  }
+
+  if (flag == Flag::NEVER_ABORTS) {
+    // The signal can't be aborted. We don't need to setup RPC
+    return js.alloc<AbortSignal>(js, /* exception */ kj::none, /* maybeReason */ kj::none, flag);
+  }
+
+  // The AbortSignalImpl will receive any remote triggerAbort requests and fulfill the promise with the reason for abort
+  auto paf = kj::newPromiseAndFulfiller<kj::Array<kj::byte>>();
+  externalHandler->setLastStream(kj::heap<AbortSignalRpcImpl>(kj::mv(paf.fulfiller)));
+
+  auto signal = js.alloc<AbortSignal>(
+      js, /* exception */ kj::none, /* maybeReason */ kj::none, flag, kj::mv(paf.promise));
+  return signal;
+}
+
+kj::Promise<void> AbortSignal::sendToRpc(
+    jsg::Lock& js, kj::Maybe<jsg::JsRef<jsg::JsValue>>& reason) {
+  jsg::Serializer ser(js);
+  KJ_IF_SOME(r, reason) {
+    ser.write(js, r.getHandle(js));
+  }
+
+  auto released = ser.release();
+
+  return sendToRpc(released.data);
+}
+
+kj::Promise<void> AbortSignal::sendToRpc(kj::ArrayPtr<kj::byte> reason) {
+  auto& ioContext = IoContext::current();
+
+  KJ_IF_SOME(outputLocks, ioContext.waitForOutputLocksIfNecessary()) {
+    co_await outputLocks;
+  }
+
+  kj::Vector<kj::Promise<void>> promises;
+  for (auto& cap: rpcClients) {
+    auto req = cap->triggerAbortRequest();
+    auto field = req.initReason();
+    field.setV8Serialized(reason);
+    promises.add(req.send().ignoreResult());
+  }
+
+  co_await kj::joinPromises(promises.releaseAsArray());
+}
+
+void AbortSignal::awaitRpcAbort(jsg::Lock& js) {
```

Comment:
> However! I think it's important that we do not do this awaitIo() unless something is actually listening on the AbortSignal. If no one actually registers to listen on the "abort" event, then we do not want the IoContext to think that it's waiting for something, because that could block hibernation inappropriately.

Here's my first crack at this problem. `awaitRpcAbort` is called when an event listener is added.  On the first listener added, I call `awaitIo` to set up a handler if a triggerAbort is received over RPC.

However this doesn't deal with `removeEventListener`. Nor something like add, remove, add. Do we care? Does there need to be some way to cancel this `awaitIo`?

Reply:
I'm also wondering if that behaviour is right. Even without registering an event handler, code can synchronously check properties like `signal.aborted`, `signal.reason` or use `singal.throwIfAborted()`